dataset-generation/spack_db/spack_db.py

```diff
@@ -0,0 +1,633 @@
+# if the program is interrupted, the saved package_hash will be 
+    # the starting point when rerun
+def save_checkpoint(package_hash): #
+    with open(CHECKPOINT_FILE, "w") as f:
+        f.write(package_hash)
+
+
+# reading file
+def readmyfile(myfile):
+    try: 
+        with open(myfile, 'r') as file:
+            # database is the spack database json, within the spack build cache
+            db = json.load(file) # 8.6 seconds to read in large json file
+
+            # returns database
+            return db
+        
+    except FileNotFoundError:
+        print(f"Error: The file '{myfile}' not found.")
+    except Exception as e:
+        print(f"Error occured in readmyfile: {e}")
+
+
+# make the spec manifest downloadable URL
+def make_spec_manifest_URL(package_hash, package_hash_value): 
+# goal is to make a URL that looks like this -> 
+# https://binaries.spack.io/develop/v3/manifests/spec/<name_of_package (not the hash)>/<name_of_package>-<version>-<hash>.spec.manifest.json
+# example URL for compiler-wrapper:
+# https://binaries.spack.io/develop/v3/manifests/spec/compiler-wrapper/compiler-wrapper-1.0-bsavlbvtqsc7yjtvka3ko3aem4wye2u3.spec.manifest.json
+    
+    myURL = 'https://binaries.spack.io/develop/v3/manifests/spec/'
+    
+    # this accesses the package name (not the package hash that we have been using as the package_hash)
+    package_name = package_hash_value['spec']['name']
+
+    # this accesses the package name version number
+    package_version = package_hash_value['spec']['version']
+    
+    # package_filename for use later in removal of placeholder directories in the filepath of the tarball
+    package_filename = (package_name + '/' + package_name + '-' + package_version 
+            + '-' + package_hash)
+
+    # this updates the URL
+    myURL += (package_filename + '.spec.manifest.json')
+    
+    # returns the URL for the spec manifest and package_filename
+    return myURL, package_filename  
+
+# automatically download contents from the URL
+def download_from_URL(theURL, package, is_spec=True): 
+
+    
+    # makes filename
+    # Example: 
+        # This -> "compiler-wrapper/compiler-wrapper-1.0-bsavlbvtqsc7yjtvka3ko3aem4wye2u3.spec.manifest.json"
+        # is turned into this -> "compiler-wrapper__compiler-wrapper-1.0-bsavlbvtqsc7yjtvka3ko3aem4wye2u3.spec.manifest.json"
+    package_name = package.replace('/','__')
+
+    # if is_spec is true, meaning the file ends with ".spec.manifest.json",
+        # then the file is not saved, but the reponse is returned to remove_lines_spec_manifest() for further manipulation
+    # if the file ends with .tar.gz
+        # then the file is saved in BINARY_CACHE_DIR
+    cache_dir = BINARY_CACHE_DIR if not is_spec else None
+
+    # full file path then is:
+        # "cache/spec_manifests/compiler-wrapper/compiler-wrapper-1.0-bsavlbvtqsc7yjtvka3ko3aem4wye2u3"
+        #cache/manifest\\compiler-wrapper-1.0-bsavlbvtqsc7yjtvka3ko3aem4wye2u3.json
```

Comment:
In terms of indentation, I'd place the `#` for all of these comments at the same level of indentation as the lines of code (then just add extra spaces to the right of the `#`). If you ran a tool like ruff on your code to follow consistent formatting, it would do the correct indentation for you throughout this entire file. Check https://docs.astral.sh/ruff/formatter/ if you want to try it out.